// @archwayhq/arch3-proto/build/archway/rewards/v1/tx.registry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageComposer = exports.load = exports.registry = void 0;
const tx_1 = require("./tx");
exports.registry = [["/archway.rewards.v1.MsgSetContractMetadata", tx_1.MsgSetContractMetadata], ["/archway.rewards.v1.MsgWithdrawRewards", tx_1.MsgWithdrawRewards], ["/archway.rewards.v1.MsgSetFlatFee", tx_1.MsgSetFlatFee], ["/archway.rewards.v1.MsgUpdateParams", tx_1.MsgUpdateParams]];
const load = (protoRegistry) => {
    exports.registry.forEach(([typeUrl, mod]) => {
        protoRegistry.register(typeUrl, mod);
    });
};
exports.load = load;
exports.MessageComposer = {
    encoded: {
        setContractMetadata(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetContractMetadata",
                value: tx_1.MsgSetContractMetadata.encode(value).finish()
            };
        },
        withdrawRewards(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgWithdrawRewards",
                value: tx_1.MsgWithdrawRewards.encode(value).finish()
            };
        },
        setFlatFee(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetFlatFee",
                value: tx_1.MsgSetFlatFee.encode(value).finish()
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgUpdateParams",
                value: tx_1.MsgUpdateParams.encode(value).finish()
            };
        }
    },
    withTypeUrl: {
        setContractMetadata(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetContractMetadata",
                value
            };
        },
        withdrawRewards(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgWithdrawRewards",
                value
            };
        },
        setFlatFee(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetFlatFee",
                value
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgUpdateParams",
                value
            };
        }
    },
    toJSON: {
        setContractMetadata(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetContractMetadata",
                value: tx_1.MsgSetContractMetadata.toJSON(value)
            };
        },
        withdrawRewards(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgWithdrawRewards",
                value: tx_1.MsgWithdrawRewards.toJSON(value)
            };
        },
        setFlatFee(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetFlatFee",
                value: tx_1.MsgSetFlatFee.toJSON(value)
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgUpdateParams",
                value: tx_1.MsgUpdateParams.toJSON(value)
            };
        }
    },
    fromJSON: {
        setContractMetadata(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetContractMetadata",
                value: tx_1.MsgSetContractMetadata.fromJSON(value)
            };
        },
        withdrawRewards(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgWithdrawRewards",
                value: tx_1.MsgWithdrawRewards.fromJSON(value)
            };
        },
        setFlatFee(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetFlatFee",
                value: tx_1.MsgSetFlatFee.fromJSON(value)
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgUpdateParams",
                value: tx_1.MsgUpdateParams.fromJSON(value)
            };
        }
    },
    fromPartial: {
        setContractMetadata(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetContractMetadata",
                value: tx_1.MsgSetContractMetadata.fromPartial(value)
            };
        },
        withdrawRewards(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgWithdrawRewards",
                value: tx_1.MsgWithdrawRewards.fromPartial(value)
            };
        },
        setFlatFee(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgSetFlatFee",
                value: tx_1.MsgSetFlatFee.fromPartial(value)
            };
        },
        updateParams(value) {
            return {
                typeUrl: "/archway.rewards.v1.MsgUpdateParams",
                value: tx_1.MsgUpdateParams.fromPartial(value)
            };
        }
    }
};
